File: Chapter 1/Section 1.4/skidesign/skidesign.cpp
#include <iostream>
#include <fstream>
#include <cmath>

/*
ID: ten.to.1
TASK: skidesign
LANG: C++
*/
//Use index based loops for better performance!

int main()
{
	std::ifstream filein("skidesign.in");
	std::ofstream fileout("skidesign.out");
	
	int N; filein >> N;
	int current_heights[N];
	int max_height = 0;
	for(int i = 0; i <= N; i++)
	{
		int inp; filein >> inp;
		current_heights[i] = inp;
		
		max_height = std::max(inp, max_height);
	}
	
	int min_cost = 999999;
	
	//Brute Force
	for(int b = 1; b <= max_height; b++)
	{
		int current_cost = 0;
		
		for(int h = 0; h < N; h++)
		{
			if(current_heights[h] < b)
			{
				current_cost += std::pow((b - current_heights[h]), 2);
			}
			else if(current_heights[h] > (b + 17))
			{
				current_cost += std::pow((current_heights[h] - (b + 17)), 2);
			}
		}
		
		min_cost = std::min(current_cost, min_cost);
	}
	
	fileout << min_cost << "\n";
}

File: Chapter 1/Section 1.4/skidesign/skidesign.py
#!/usr/bin/python

"""
ID: ten.to.1
TASK: skidesign
LANG: PYTHON3
"""

filein  = open("skidesign.in", "r")
fileout = open("skidesign.out", "w")

filein.readline()
hill_heigths = sorted(list(map(int, filein.read().split())))
changes      = []

print(hill_heigths)

while(True):
	max_height    = max(hill_heigths)
	min_height    = min(hill_heigths)
	displacement  = max_height - min_height
	change_needed = int((displacement - 17) / 2)
	
	if(change_needed == 0):
		break
	
	#print(min(hill_heigths), max(hill_heigths))
	#print("\t", displacement)
	#print(hill_heigths)
	
	changes.append(change_needed)
	changes.append(change_needed)
	
	hill_heigths[hill_heigths.index(max_height)] = hill_heigths[hill_heigths.index(max_height)] - change_needed
	hill_heigths[hill_heigths.index(min_height)] = hill_heigths[hill_heigths.index(min_height)] + change_needed

print(changes)

cost = 0
for n in changes:
	cost += n**2

fileout.write(str(cost) + '\n')

